src/metabase/data_apps/models.clj

```diff
@@ -0,0 +1,128 @@
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(methodical/defmethod t2/table-name :model/App           [_model] :app)
+(methodical/defmethod t2/table-name :model/AppDefinition [_model] :app_definition)
+(methodical/defmethod t2/table-name :model/AppRelease    [_model] :app_release)
+
+(doseq [model [:model/App
+               :model/AppDefinition
+               :model/AppRelease]]
+  (derive model :metabase/model))
+
+(derive :model/App :hook/timestamped?)
+(derive :model/AppDefinition :hook/created-at-timestamped?)
+(derive :model/AppDefinition :hook/entity-id)
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                         :model/App                                             ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(t2/deftransforms :model/App
+  {:config mi/transform-json})
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                     :model/AppDefinition                                       ;;
+;;-----------------------------------------------------------------------------------------------;;
+
+(mr/def ::AppDefinitionConfig
+  [:maybe [:map {:closed true}
+           [:version pos-int?]]])
+
+(defn- validate-app-definition
+  "Validate an AppDefinition."
+  [app-definition]
+  (mu/validate-throw ::AppDefinitionConfig (:config app-definition)))
+
+(defn- next-version
+  [app-id]
+  (or (t2/select-one-fn :version :model/AppDefinition :app_id app-id {:order-by [[:version :desc]]
+                                                                      :limit 1})
+      0))
+
+(t2/define-before-insert :model/AppDefinition
+  [instance]
+  (validate-app-definition instance)
+  (assoc instance :version (next-version (:app_id instance))))
+
+(t2/define-before-update :model/AppDefinition
+  [instance]
+  ;; AppDefinition is append-only, so prevent updates
+  (throw (ex-info "AppDefinition is append-only and cannot be updated"
+                  {:status-code 400
+                   :changes     (t2/changes instance)})))
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                     :model/AppRelease                                       ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(defn- ensure-single-active-release!
+  "Ensure at most one active publication per app_id."
+  [instance]
+  (when (:active instance)
+    (t2/update! :model/AppRelease
+                {:app_id (:app_id instance)
+                 :active true}
+                {:active false}))
+  instance)
+
+(t2/define-before-insert :model/AppRelease
```

Comment:
Not a fan of this hook, as there's action-at-a-distance assuming that we always insert active releases.